### PR TITLE
bug fix in checkodesol

### LINKED ISSUE #15996: Dsolve: undetermined coefficient fails but variation of parameters returns
There are several different possible issues here. This example comes from the tests:
[https://github.com/sympy/sympy/blob/38174f7373a86ac92527dd7120cd92c668a38560/sympy/solvers/tests/test_ode.py#L2236](https://github.com/sympy/sympy/blob/38174f7373a86ac92527dd7120cd92c668a38560/sympy/solvers/tests/test_ode.py#L2236)

1. The ODE below gives NotImplementedError under `dsolve` even though variation of parameters can find a solution.
2. The ODE matches undetermined coefficients which then fails to solve it (either it shouldn't match or it should succeed).
3. I think there might be many cases like this in `dsolve`: because one solver has matched the ODE other solvers cannot attempt to solve it (the top-level algorithm of dsolve needs to change - probably needs discussing in its own issue).
4. The solution returned by variation of parameters tests as being incorrect under `checkodesol` so either a bug in checkodesol or variation of parameters (I'm not sure as I haven't verified the solution myself).

```julia
In [15]: eq = f(x).diff(x, 5) + 2*f(x).diff(x, 3) + f(x).diff(x) - 2*x - exp(I*x)                                                                             

In [20]: eq                                                                                                                                                   
Out[20]: 
                             3           5      
        ⅈ⋅x   d             d           d       
-2⋅x - ℯ    + ──(f(x)) + 2⋅───(f(x)) + ───(f(x))
              dx             3           5      
                           dx          dx       

In [16]: dsolve(eq, hint='nth_linear_constant_coeff_variation_of_parameters')                                                                                 
Out[16]: 
                 ⎛      2     ⎛        ⅈ⋅x      -ⅈ⋅x⎞      2⋅ⅈ⋅x                       ⎞          ⎛        2     ⎛          ⅈ⋅x        -ⅈ⋅x⎞        2⋅ⅈ⋅x   
             2   ⎜     x      ⎜     3⋅ℯ      3⋅ℯ    ⎟   5⋅ℯ             ⅈ⋅x        -ⅈ⋅x⎟          ⎜     ⅈ⋅x      ⎜     3⋅ⅈ⋅ℯ      3⋅ⅈ⋅ℯ    ⎟   5⋅ⅈ⋅ℯ        
f(x) = C₁ + x  + ⎜C₂ - ── + x⋅⎜C₃ + ────── + ───────⎟ + ──────── + 2⋅ⅈ⋅ℯ    - 2⋅ⅈ⋅ℯ    ⎟⋅sin(x) + ⎜C₄ + ──── + x⋅⎜C₅ + ──────── - ─────────⎟ + ────────── - 
                 ⎝     8      ⎝       2         2   ⎠      16                          ⎠          ⎝      8       ⎝        2           2    ⎠       16       

                ⎞                
   ⅈ⋅x      -ⅈ⋅x⎟             ⅈ⋅x
2⋅ℯ    - 2⋅ℯ    ⎟⋅cos(x) - ⅈ⋅ℯ   
                ⎠                

In [17]: checkodesol(eq, _)                                                                                                                                   
Out[17]: 
(False,
 (3*exp(I*x)*sin(x) + 3*I*exp(I*x)*cos(x) - 2*exp(I*x) + 2*I*sin(x) + 2*cos(x))*exp(I*x) + 3*sin(x) - 3*I*cos(x))

In [18]: dsolve(eq)                                                                                                                                           
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-18-42dc4d2e7b25> in <module>
----> 1 dsolve(eq)

~/current/sympy/sympy/sympy/solvers/ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    663             # The key 'hint' stores the hint needed to be solved for.
    664             hint = hints['hint']
--> 665             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    666 
    667 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~/current/sympy/sympy/sympy/solvers/ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    688         # attempt to solve for func, and apply any other hint specific
    689         # simplifications
--> 690         sols = solvefunc(eq, func, order, match)
    691         if isinstance(sols, Expr):
    692             rv =  odesimp(sols, func, order, cons(sols), hint)

~/current/sympy/sympy/sympy/solvers/ode.py in ode_nth_linear_constant_coeff_undetermined_coefficients(eq, func, order, match)
   5071         returns='both')
   5072     match.update(gensol)
-> 5073     return _solve_undetermined_coefficients(eq, func, order, match)
   5074 
   5075 

~/current/sympy/sympy/sympy/solvers/ode.py in _solve_undetermined_coefficients(eq, func, order, match)
   5172             "Could not solve `%s` using the "
   5173             "method of undetermined coefficients "
-> 5174             "(unable to solve for coefficients)." % eq)
   5175 
   5176     psol = trialfunc.subs(coeffvals)

NotImplementedError: Could not solve `-2*x - exp(I*x) + Derivative(f(x), x) + 2*Derivative(f(x), (x, 3)) + Derivative(f(x), (x, 5))` using the method of undetermined coefficients (unable to solve for coefficients).

In [19]: classify_ode(eq)                                                                                                                                     
Out[19]: 
('nth_linear_constant_coeff_undetermined_coefficients',
 'nth_linear_constant_coeff_variation_of_parameters',
 'nth_linear_constant_coeff_variation_of_parameters_Integral')
```

### PR DESCRIPTION
**Pull request by Mohitbalwani26**
#### References to other Issues or PRs

Fixes #15996 

#### Brief description of what is fixed or changed
`rewrite(exp)`  is used inside `checkodesol` because the solution returned by `variation of parameters` was correct but `checkodesol` couldn't simplify

#### Release Notes

* solvers
  *  Fixed `checkodesol` for some cases when the solution returned by `variation of parameters` was correct but `chekodesol` couldn't simplify it.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
This will need tests:
1. A test for `checkodesol`.
2. A test for the ODE in variation of parameters.

**Comment by Mohitbalwani26** [comment]
> This will need tests:
> 
> 1. A test for `checkodesol`.
> 2. A test for the ODE in variation of parameters.

test for variation of parameters will also be test for `checkodesol` right?

File: sympy/solvers/tests/test_ode.py [hunk 1]
```diff
@@ -2333,13 +2333,13 @@ def test_nth_linear_constant_coeff_variation_of_parameters_simplify_False():
...
-    assert not zero.is_zero and zero.rewrite(exp).simplify() == 0
+    assert checkodesol(eq, sol_simp, order=5, solve_for_func=False)[0]
+    # assert not zero.is_zero and zero.rewrite(exp).simplify() == 0
     # \-----------
     (t, zero) = checkodesol(eq, sol_nsimp, order=5, solve_for_func=False)
     # if this fails because zero.is_zero, replace this block with
-    # assert checkodesol(eq, sol_simp, order=5, solve_for_func=False)[0]
-    assert zero == 0
+    assert checkodesol(eq, sol_simp, order=5, solve_for_func=False)[0]
+    # assert zero == 0
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_ode.py]: If `checkodesol` now works for this case we can clean this up and remove all of those comments.

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: I'd prefer this though:
```
assert checkodesol(...) == (True, 0)
```

**Comment by Mohitbalwani26** [comment]
Yes, i will do this once the travis passes.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin does this look good?

**Comment by oscarbenjamin** [comment]
Yes, looks good

**Pull request merged**